internals/idkey/idkey.go

```diff
@@ -0,0 +1,265 @@
+	err = expectPermission(k.keyDir, 0o700)
+	if err != nil {
+		return err
+	}
+	pemPath := filepath.Join(k.keyDir, identityKeyFile)
+	err = expectPermission(pemPath, 0o600)
+	if err != nil {
+		return err
+	}
+	// Load the key.
+	pemData, err := os.ReadFile(pemPath)
+	if err != nil {
+		return err
+	}
+	for {
+		var block *pem.Block
+		block, pemData = pem.Decode(pemData)
+		if block == nil {
+			break
+		}
+		switch block.Type {
+		case "PRIVATE KEY":
+			k.key, err = x509.ParsePKCS8PrivateKey(block.Bytes)
+			if err != nil {
+				return err
+			}
+		default:
+			return fmt.Errorf("unknown PEM block %q in file %q", block.Type, pemPath)
+		}
+	}
+	if k.key == nil {
+		return fmt.Errorf("empty PEM file %q", pemPath)
+	}
+	return nil
+}
+
+// save saves the private identity key to storage.
+func (k *IDKey) save() error {
+	exists, err := pathExists(k.keyDir)
+	if err != nil {
+		return fmt.Errorf("directory %q is not accessible", k.keyDir)
+	}
+	if exists {
+		err = expectPermission(k.keyDir, 0o700)
+		if err != nil {
+			return err
+		}
+	} else {
+		// Create the leaf directory node with 0700 permissions.
+		err = os.Mkdir(k.keyDir, 0o700)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Create the new private identity file.
+	pemPath := filepath.Join(k.keyDir, identityKeyFile)
+	pemFile, err := os.OpenFile(pemPath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0o600)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		err = errors.Join(err, pemFile.Sync())
+		err = errors.Join(err, pemFile.Close())
+	}()
+	keyBytes, err := x509.MarshalPKCS8PrivateKey(k.key)
+	if err != nil {
+		return err
+	}
+	pemPrivateBlock := &pem.Block{
+		Type:  "PRIVATE KEY",
+		Bytes: keyBytes,
+	}
+	var pemBuffer bytes.Buffer
+	if err := pem.Encode(&pemBuffer, pemPrivateBlock); err != nil {
+		return err
+	}
+	if _, err = pemFile.Write(pemBuffer.Bytes()); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Fingerprint returns the identity fingerprint. This is the SHA512/384 hash
+// of the public key, encoded in base32 (without padding). This is a
+// convenient shorthand form of the identity that can be used to identify
+// a specfic machine or device.
+func (k *IDKey) Fingerprint() string {
+	publicBytes := k.Public().(ed25519.PublicKey)
+	hashBytes := sha512.Sum384(publicBytes)
+	return base32.StdEncoding.WithPadding(base32.NoPadding).EncodeToString(hashBytes[:])
+}
+
+// expectPermission return an error if the specified directory or file
+// path is not matching the supplied permissions.
+func expectPermission(path string, perm fs.FileMode) error {
+	info, err := os.Stat(path)
+	if err != nil {
+		return err
+	}
+	actualPerm := info.Mode().Perm()
+	if actualPerm != perm {
+		return fmt.Errorf("expected permission 0o%o (got 0o%o) for %q", perm, actualPerm, path)
+	}
+	return nil
+}
+
+// pathExists returns true of the path exists, false if it does not. If the
+// operation reports an unrelated error, the error is returned.
+func pathExists(path string) (bool, error) {
```

Comment:
nit: we have `osutil.ExistsIsDir`, if we want to get rid of this

Reply:
I reckon we should leave it. That `ExistsIsDir` function has a confusing name and odd return values -- it tries to do too much IMO.

Reply:
Resolving.